### PR TITLE
Unsubscribe from Subscription via signed url

### PR DESCRIPTION
**Pull request by garyhtou**
Users can unsubscribe from a `Subscription` using a signed URL. This URL will be sent in `Digest` mailers. It's a simple `GET` to `/subscriptions/:id/unsubscribe`, where `:id` is a [signed id](https://api.rubyonrails.org/classes/ActiveRecord/SignedId.html) with an `unsubscribe` purpose.

Upon visiting an unsubscribe URL, you will see the following:
image

Unsubscribe URLs expire after 2 months.
image

If the signature in the signed id is invalid, the view will show it as an expired URL.

---

**Example URL:**
```
http://localhost:3000/subscriptions/eyJfcmFpbHMiOnsibWVzc2FnZSI6Ik16QT0iLCJleHAiOiIyMDIzLTA5LTI2VDA1OjM3OjQ4LjQ5M1oiLCJwdXIiOiJoYWNrYXRob24vc3Vic2NyaXB0aW9uL3Vuc3Vic2NyaWJlIn19--b3e99639b45f15025099b3241ea287f2f921e3e26b1a168265f0c6578ca9c48b/unsubscribe
```

### CODE REVIEW

File: app/controllers/hackathon/subscriptions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class Hackathon::SubscriptionsController < ApplicationController
+  skip_before_action :redirect_if_unauthenticated, only: [:unsubscribe]
+
+  def unsubscribe
```
Comment (northeastprince) [inline app/controllers/hackathon/subscriptions_controller.rb:4]: How about `edit` or in a `Subscriptions::RemovalsController`?

**Review by northeastprince — CHANGES_REQUESTED** [review]

File: test/controllers/hackathon/subscriptions_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
+require "test_helper"
+
+class Hackathon::SubscriptionsControllerTest < ActionDispatch::IntegrationTest
```
Comment (northeastprince) [inline test/controllers/hackathon/subscriptions_controller_test.rb:3]: Maybe move this to a system test for consistency?

File: test/controllers/hackathon/subscriptions_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  setup do
+    @subscription = hackathon_subscriptions(:gary_seattle)
+  end
+
+  test "should unsubscribe" do
```
Comment (northeastprince) [inline test/controllers/hackathon/subscriptions_controller_test.rb:8]:
```suggestion
  test "unsubscribing with a valid link" do
```

File: test/controllers/hackathon/subscriptions_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    get @subscription.unsubscribe_url
+    assert_response :success
+
+    assert_match "You have unsubscribed", @response.body
+    assert_match @subscription.location, @response.body
+  end
+
+  test "should expire" do
```
Comment (northeastprince) [inline test/controllers/hackathon/subscriptions_controller_test.rb:16]:
```suggestion
  test "unsubscribing with an expired link" do
```

File: test/controllers/hackathon/subscriptions_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+    travel_to expiration_at do
+      get unsub_url
+      assert_response :success
+
+      assert_match "expired", @response.body
+    end
+  end
+
+  test "should handle invalid signature" do
```
Comment (northeastprince) [inline test/controllers/hackathon/subscriptions_controller_test.rb:28]:
```suggestion
  test "unsubscribing with an invalid link" do
```

File: app/models/hackathon/subscription/status.rb [hunk 6]
```diff
@@ -9,6 +9,15 @@ module Hackathon::Subscription::Status
     scope :active_for, ->(user) { active.where(subscriber: user) }
   end
 
+  # Unsubscribe URLs must be valid for at least 30 days.
+  # https://www.ftc.gov/business-guidance/resources/can-spam-act-compliance-guide-business
+  UNSUBSCRIBE_EXPIRATION = 2.months
+
+  def unsubscribe_url
```
Comment (northeastprince) [inline app/models/hackathon/subscription/status.rb:16]: I don't think we should have a helper like this in the model.

**Comment by garyhtou** [comment]
Since we ended up building out [`Digest`](https://github.com/hackclub/hackathons-backend/issues/39), each email will be contain **multiple** subscriptions. It's a little weird to have multiple unsubscribe links in an email. So I think this PR might be scrapped in favor of a "manage subscriptions" page.

The "manage subscriptions" will likely require some sort of ephemeral authentication provided via a signed unsubscribe url

**Comment by garyhtou** [comment]
Closing in favor of https://github.com/hackclub/hackathons-backend/pull/85

**Pull request closed without merging**